8193445: JavaFX CSS is applied redundantly leading to significant performance degradation

aghaisas opened the pull request:
**Issue :**
https://bugs.openjdk.java.net/browse/JDK-8193445

**Background :**
The CSS performance improvement done in [JDK-8151756](https://bugs.openjdk.java.net/browse/JDK-8151756) had to be backed out due to functional regressions reported in [JDK-8185709](https://bugs.openjdk.java.net/browse/JDK-8185709), [JDK-8183100](https://bugs.openjdk.java.net/browse/JDK-8183100) and [JDK-8168951](https://bugs.openjdk.java.net/browse/JDK-8168951).
Refer to [JDK-8183100](https://bugs.openjdk.java.net/browse/JDK-8183100) for more details on this backout. 

**Description :**
This PR reintroduces the CSS performance improvement fix done in [JDK-8151756](https://bugs.openjdk.java.net/browse/JDK-8151756) while addressing the functional regressions that were reported in [JDK-8185709](https://bugs.openjdk.java.net/browse/JDK-8185709), [JDK-8183100](https://bugs.openjdk.java.net/browse/JDK-8183100) and [JDK-8168951](https://bugs.openjdk.java.net/browse/JDK-8168951).
For ease of review, I have made two separate commits -
1) [Commit 1](https://github.com/openjdk/jfx/pull/34/commits/d964675ebc2a42f2fd6928b773819502683f2334) - Reintroduces the CSS performance improvement fix done in [JDK-8151756](https://bugs.openjdk.java.net/browse/JDK-8151756) - most of the patch applied cleanly.
2) [Commit 2 ](https://github.com/openjdk/jfx/pull/34/commits/12ea8220a730ff8d98d520ce870691d54f0de00f)- fixes the functional regressions keeping performance improvement intact + adds a system test

**Root Cause :**
CSS performance improvement fix proposed in [JDK-8151756 ](https://bugs.openjdk.java.net/browse/JDK-8151756)correctly avoids the redundant CSS reapplication to children of a Parent. 
What was missed earlier in [JDK-8151756 ](https://bugs.openjdk.java.net/browse/JDK-8151756) fix : "CSS reapplication to the Parent itself”. 
This missing piece was the root cause of all functional regressions reported against [JDK-8151756](https://bugs.openjdk.java.net/browse/JDK-8151756)

**Fix :**
Fixed the identified root cause. See commit 2.

**Testing :**
1. All passing unit tests continue to pass
2. New system test (based on [JDK-8209830](https://bugs.openjdk.java.net/browse/JDK-8209830)) added in this PR - fails before this fix and passes after the fix
3. System test JDK8183100Test continues to pass
4. All test cases attached to regressions [JDK-8185709](https://bugs.openjdk.java.net/browse/JDK-8185709), [JDK-8183100](https://bugs.openjdk.java.net/browse/JDK-8183100) and [JDK-8168951](https://bugs.openjdk.java.net/browse/JDK-8168951) pass with this fix

In addition, testing by community with specific CSS performance / functionality will be helpful.
<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
## Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

## Issue
[JDK-8193445](https://bugs.openjdk.java.net/browse/JDK-8193445): JavaFX CSS is applied redundantly leading to significant performance degradation


## Approvers
 * Kevin Rushforth ([kcr](@kevinrushforth) - **Reviewer**)
 * David Grieve ([dgrieve](@dsgrieve) - **Reviewer**) **Note!** Review applies to 4dade6e5e3cf3a62c8e63eb12a9e732ee4d6d35e

aghaisas commented:
Reviewers: @kevinrushforth and @arapte

aghaisas commented:
It will be helpful if I can get some feedback from community with additional testing apart from what is mentioned in the description.

dsgrieve commented:
Has this been checked with SubScene? Two cases would be 1) SubScene inheriting a style from its parent, and 2) behavior of a parent in the SubScene itself. I don't expect that this would be an issue, but there is some special handling of CSS in SubScene IIRC.

DeanWookey commented:
Having looked at this issue previously (https://mail.openjdk.java.net/pipermail/openjfx-dev/2018-November/022842.html and https://github.com/DeanWookey/openjdk-jfx/commit/65a1ed82bce262294f1969e9a12e1126ec8a1ec6), I'm a bit confused. 

Doesn't commit https://github.com/openjdk/jfx/pull/34/commits/12ea8220a730ff8d98d520ce870691d54f0de00f essentially add another reapplyCSS() above the scenesChanged call on line 1074? I'm guessing this works because reapplyCss() (different from reapplyCSS()) sets cssFlag to UPDATE, which then means subsequent calls to reapplyCSS() don't call reapplyCss()?

Does this leave the whole tree with the CssFlags.REAPPLY set instead of CssFlags.UPDATE as they would be without these changes? I'm not sure about the impact of that.

I'm also curious whether commit 1 is even necessary with commit 2.

kevinrushforth commented:
The above raises good questions about the effect of adding a call to `reapplyCSS` when `scenesChanged` is called. This will need a careful analysis.

Regarding whether commit 2 by itself would be sufficient, my initial guess was that it wouldn't be. However, I tested it, and it does in fact improve performance of the test case. Very interesting.

kevinrushforth commented:
Looking at this a little further, the change from commit 2, to call `reapplyCSS` at the beginning of `scenesChanged` really does two things:

1. Calling `reapplyCSS` right at the beginning of `scenesChanged` flips the order of CSS reapplication from the existing "bottom up" order to a "top down" order. Note that `scenesChanged` calls `invalidatedScenes` recursively (via `setScenes`) to traverse the scene graph.

2. It effectively undoes the rest of the changes, which were intended to avoid calling `reapplyCSS` in cases where it is redundant. As long as the performance improvement is preserved, this seems fine.

The first of the above appears to be both necessary and sufficient to both maintain correctness while still avoiding the n^2 application of CSS in the case of adding a deep scene graph. If this is in fact the case, then I think the fix can be simplified even further by moving the following call from its current location (after `updateTreeShowing`) to right before the call to `scenesChanged`:

```
    if (sceneChanged) reapplyCSS();
```

in which case the changes to `Node::scenesChanged` and `Parent::scenesChanged` can be reverted

The following patch, by itself (with nothing from commit 1) might be sufficient:

```
--- a/modules/javafx.graphics/src/main/java/javafx/scene/Node.java
+++ b/modules/javafx.graphics/src/main/java/javafx/scene/Node.java
@@ -1069,6 +1069,7 @@ public abstract class Node implements EventTarget, Styleable {
             focusSetDirty(oldScene);
             focusSetDirty(newScene);
         }
+        if (sceneChanged) reapplyCSS(); // Note this can be moved inside the previous 'if' block
         scenesChanged(newScene, newSubScene, oldScene, oldSubScene);

         // isTreeShowing needs to take into account of Window's showing
@@ -1091,8 +1092,6 @@ public abstract class Node implements EventTarget, Styleable {
         }
         updateTreeShowing();

-        if (sceneChanged) reapplyCSS();
-
         if (sceneChanged && !isDirtyEmpty()) {
             //Note: no need to remove from scene's dirty list
             //Scene's is checking if the node's scene is correct
```

We will need to do a more complete evaluation as to whether this inversion of calling reapplyCSS on the parent before its children has any side effects, and also whether there are any of the performance improvements from the original fix that aren't covered here.

DeanWookey commented:
I think inverting the call is fine. That's what I did in my fix (https://github.com/DeanWookey/openjdk-jfx/commit/65a1ed82bce262294f1969e9a12e1126ec8a1ec6) and we've been testing that out thoroughly for over a year.

It's as if you are adding nodes 1 by 1 to the scene graph, something we know works and is fast. My change tries to emulate that more accurately to avoid side effects. Theoretically, we should be able to do better when many nodes are added at once because we have all the information upfront. 

The one side effect I can see by only applying commit 2 is that the first call of reapplyCSS() calls reapplyCss on every node in the tree and that sets the cssFlag = CssFlags.UPDATE;. The subsequent calls will hit this in reapplyCSS():

```
        if (cssFlag == CssFlags.UPDATE) {
            cssFlag = CssFlags.REAPPLY;
            notifyParentsOfInvalidatedCSS();
            return;
        }
```
and return without doing all the unnecessary work. As a result however, instead of leaving with cssFlag = CssFlags.UPDATE, all the nodes leave with CssFlags.REAPPLY. That might cause an unnecessary css pass later? 

Doing it in the order it happens now, that check for the update flag shouldn't be true because its bottom up.

dsgrieve commented:
I think the order of calls in invalidatedScenes is correct. The scene graph plumbing needs to be done before CSS is applied since a node's style can depend on the styles of its parent.

The issue is really calling reapplyCss(). In the case where a child is added to a parent, the child's parent property is invalidated which should call reapplyCSS(), which should (in turn) call reapplyCss(). Next the invalidatedScenes call happens and the whole dance starts again. (See Parent children.onChanged). 

Perhaps short-circuiting the call to reapplyCss() from the reapplyCSS() method is the thing to do. Since CSS is reapplied from the top down, my parent's cssFlag should be CLEAN before reapplyCss() method is called from reapplyCSS().

It has been a long time since I've been in this code. I'm going to look at this some more.

kevinrushforth commented on tests/system/src/test/java/test/robot/javafx/scene/CSSPerf_JDK8193445Test.java:
```diff
@@ -0,0 +1,96 @@
+ * questions.
+ */
+
+package test.robot.javafx.scene;
```
There is no need for this test to require robot. I recommend moving it to `test.javafx.scene` (and not inherit from `VisualTestBase`).

kevinrushforth commented on tests/system/src/test/java/test/robot/javafx/scene/CSSPerf_JDK8193445Test.java:
```diff
@@ -0,0 +1,96 @@
+ * in performance, if any.
+ */
+
+public class CSSPerf_JDK8193445Test extends VisualTestBase {
```
We have moved away from putting the bug ID in the test class name, so I recommend renaming it.

kevinrushforth commented on tests/system/src/test/java/test/robot/javafx/scene/CSSPerf_JDK8193445Test.java:
```diff
@@ -0,0 +1,96 @@
+            long startTime = System.currentTimeMillis();
+
+            HBox hbox = new HBox();
+            for (int i = 0; i < 300; i++) {
```
In my testing on various machines, the bug is more pronounced, and less prone to system differences with `500` nodes instead of `300`.

kevinrushforth commented on tests/system/src/test/java/test/robot/javafx/scene/CSSPerf_JDK8193445Test.java:
```diff
@@ -0,0 +1,96 @@
+
+        // NOTE : 400 mSec is not a benchmark value
+        // It is good enough to catch the regression in performance, if any
+        assertTrue("Time to add 300 Nodes is more than 400 mSec", mSec < 400);
```
If you increase the number of nodes to `500` then I recommend bumping the time threshold to `800` msec in case it is run on a very slow system.

aghaisas commented:
> I think inverting the call is fine. That's what I did in my fix ([DeanWookey/openjdk-jfx@65a1ed8](https://github.com/DeanWookey/openjdk-jfx/commit/65a1ed82bce262294f1969e9a12e1126ec8a1ec6)) and we've been testing that out thoroughly for over a year.
> 
> It's as if you are adding nodes 1 by 1 to the scene graph, something we know works and is fast. My change tries to emulate that more accurately to avoid side effects. Theoretically, we should be able to do better when many nodes are added at once because we have all the information upfront.
> 
> The one side effect I can see by only applying commit 2 is that the first call of reapplyCSS() calls reapplyCss on every node in the tree and that sets the cssFlag = CssFlags.UPDATE;. The subsequent calls will hit this in reapplyCSS():
> 
> ```
>         if (cssFlag == CssFlags.UPDATE) {
>             cssFlag = CssFlags.REAPPLY;
>             notifyParentsOfInvalidatedCSS();
>             return;
>         }
> ```
> 
> and return without doing all the unnecessary work. As a result however, instead of leaving with cssFlag = CssFlags.UPDATE, all the nodes leave with CssFlags.REAPPLY. That might cause an unnecessary css pass later?
> 
> Doing it in the order it happens now, that check for the update flag shouldn't be true because its bottom up.

It is a good observation about cssFlag. I have not seen any side effect with the limited testing that I have done. It may be possible that the "unnecessary css pass later" scenario is not covered by the test cases that we have.

aghaisas commented:
> Perhaps short-circuiting the call to reapplyCss() from the reapplyCSS() method is the thing to do.

This comment from @dsgrieve got me interested. I checked the test code JDK-8151756 with cssFlags logged, it became evident that the cssFlag gets set to DIRTY_BRANCH for every parent and reapplyCss() gets invoked for each of the children. This is the exact redundant processing.


Test from JDK-8151756 with additional one level of Node hierarchy.

Parent1<--Parent2<--Parent3<--Rectangle (leaf child)

Log from test program ----
Parent 1 : VBox@1d9e402b
Parent 2 : VBox@4cc2dcce
Parent 3 : VBox@4cc2dcce
Rectangle 

**REAPPLY_CSS called for : VBox@1d9e402b ----- CssFlags.CLEAN
REAPPLY_CSS called for : Rectangle[...] ----- CssFlags.CLEAN**
reapplyCss called for : Rectangle[...] ----- CssFlags.CLEAN
**REAPPLY_CSS called for : VBox@19234c0d ----- CssFlags.DIRTY_BRANCH**
reapplyCss called for : VBox@19234c0d ----- CssFlags.DIRTY_BRANCH
reapplyCss called for : Rectangle[...] ----- CssFlags.CLEAN
**REAPPLY_CSS called for : VBox@4cc2dcce ----- CssFlags.DIRTY_BRANCH**
reapplyCss called for : VBox@4cc2dcce ----- CssFlags.DIRTY_BRANCH
reapplyCss called for : VBox@19234c0d ----- CssFlags.UPDATE
reapplyCss called for : Rectangle[...] ----- CssFlags.CLEAN
**REAPPLY_CSS called for : VBox@1d9e402b ----- CssFlags.DIRTY_BRANCH**
reapplyCss called for : VBox@1d9e402b ----- CssFlags.DIRTY_BRANCH
reapplyCss called for : VBox@4cc2dcce ----- CssFlags.UPDATE
reapplyCss called for : VBox@19234c0d ----- CssFlags.UPDATE
reapplyCss called for : Rectangle[...] ----- CssFlags.CLEAN


Proposed New Fix :
-------------------
I added a simple check to avoid reapplyCss() call for each Node with DIRTY_BRANCH cssFlag. Here is the patch -

```
diff --git a/modules/javafx.graphics/src/main/java/javafx/scene/Node.java b/modules/javafx.graphics/src/main/java/javafx/scene/Node.java
index 877e0fd6c8..8606dfb575 100644
--- a/modules/javafx.graphics/src/main/java/javafx/scene/Node.java
+++ b/modules/javafx.graphics/src/main/java/javafx/scene/Node.java
@@ -9416,7 +9416,7 @@ public abstract class Node implements EventTarget, Styleable {
         if (cssFlag == CssFlags.REAPPLY) return;
 
         // RT-36838 - don't reapply CSS in the middle of an update
-        if (cssFlag == CssFlags.UPDATE) {
+        if (cssFlag == CssFlags.UPDATE || cssFlag == CssFlags.DIRTY_BRANCH) {
             cssFlag = CssFlags.REAPPLY;
             notifyParentsOfInvalidatedCSS();
             return;

```
With this fix -
Log from test program ----
Parent 1 : VBox@36d24c70
Parent 2 : VBox@35af5cea
Parent 3 : VBox@35af5cea
Rectangle

**REAPPLY_CSS called for : VBox@36d24c70 ----- CssFlags.CLEAN**
**REAPPLY_CSS called for : Rectangle[...] ----- CssFlags.CLEAN**
reapplyCss called for : Rectangle[...] ----- CssFlags.CLEAN
**REAPPLY_CSS called for : VBox@5d4b6983 ----- CssFlags.DIRTY_BRANCH
REAPPLY_CSS called for : VBox@35af5cea ----- CssFlags.DIRTY_BRANCH
REAPPLY_CSS called for : VBox@36d24c70 ----- CssFlags.DIRTY_BRANCH**
reapplyCss called for : VBox@36d24c70 ----- CssFlags.REAPPLY
reapplyCss called for : VBox@35af5cea ----- CssFlags.REAPPLY
reapplyCss called for : VBox@5d4b6983 ----- CssFlags.REAPPLY
reapplyCss called for : Rectangle[...] ----- CssFlags.CLEAN


I verified that all graphics/controls unit tests & all system tests pass with this fix.
I launched and played with Ensemble app. I did not see any visible artifacts.

dsgrieve commented:
@aghaisas You can avoid the call to notifyParentsOfInvalidatedCSS in the case where the flag is DIRTY_BRANCH. 

I like the looks of this. From the 10,000 foot view, when a node's parent changes, or a node's scene changes, CSS should be reapplied. This is exactly what 'if (sceneChanged) reapplyCSS()' says, and what happens in parent property's invalidated method. All of the optimizations (do I _really_ need to reapply css?) happen elsewhere, so I like this solution better than passing a boolean around (the original patch).

aghaisas commented:
Thanks @dsgrieve for having a look. I have updated the PR as suggested to avoid call to notifyParentsOfInvalidatedCSS in case the flag is DIRTY_BRANCH.
Also, I have modified the system test as suggested by @kevinrushforth.

Kindly review.

Limited testing shows this fix holds up good.

dsgrieve commented:
Trying to run this, but have to build on Windows. Ugh!

aghaisas commented:
Request to @DeanWookey, @tomsontom, @swpalmer - can you please confirm if this fix helps your application or tests?

Groostav commented:
Does the performance problem addressed here relate to reducing heap space allocations? My profiler pointed me at this issue, and around the same time I'm seeing `OutOfMemoryError`s. I'm wondering if (or rather, _hoping_) they're the same problem.

aghaisas commented:
I did a quick check w.r.t. memory using profiler. This patch significantly reduces the heap allocations as well.
There is a win on both - performance & memory front.

DJViking commented:
I am curious. Will∕Could this CSS performance improvement be backported to JavaFX 11? The bug report says only that it will be fixed in JavaFX 14.

aghaisas commented:
/integrate

buedi commented:
> Looks good.
Do you think this fix will be backported to JDK8\JavaFX8?

kevinrushforth commented:
> Do you think this fix will be backported to JDK8\JavaFX8?

This has already been backported to FX 8 for Oracle's JDK 8u251.

DJViking commented:
> > Do you think this fix will be backported to JDK8\JavaFX8?
> 
> This has already been backported to FX 8 for Oracle's JDK 8u251.

Will JavaFX 11 get the same treatment?

kevinrushforth commented:
Yes, this would be an excellent candidate to backport to JavaFX 11 (and one I was going to recommend).
